BZ 64265 Fix and simplify weak ETag matching

stokito opened the pull request:
The If-None-Match header may have multiple ETags https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.26
Even if Tomcat returns a single ETag clients may want to send few of them.
But Tomcat also makes weak matching while resource ETag is a weak by default. This functionality was implemented in b26ad77f659de34d49dda5a248566a38383cec0e but only for `If-Match` header while `If-None-Match` header remains unfixed. 

The PR contains a fix of weak matching for `If-None-Match`.

Also since ETags are always quoted the easiest way to check is just to check a substring.
This will reduce amount of memory consumed by full ETag parsing.

A commands to check:

    curl -vv http://localhost:8080/tomcat.png -H 'If-None-Match: W/"5103-1595246577000", "etag2"'
    curl -vv http://localhost:8080/tomcat.png -H 'If-None-Match: "5103-1595246577000", "etag2"'

Both returns 304 Not Modified.

reschke commented:
The suggested change will cause incorrect behaviour, for instance

    "foobar"

would match

    W/"foobar"

That said, the existing code is broken as well, as it will fail if an ETag contains a comma.

stokito commented:
Thank you, good points! I checked spec https://tools.ietf.org/html/rfc7232#section-2.3.2 and it looks like we still fine here because Tomcat in fact should do a weak comparison.

|If-None-Match | Real resource's ETag | Match              |
|--------------------|-----------------------------|---------------------|
|`W/"1"`            | `W/"1"`                       | Yes                   |
|`"1"`                | `W/"1"`                       | No.                    |
|`W/"1"`            | `"1"`                           | Yes, as in spec |
|`"1"`                | `"1"`                           | Yes                    |
|`W/"1", "1"`     | `"1"`                           | Yes                    |
|`W/"1", "2"`     | `W/"2"`                       | No                     |

reschke commented:
You are citing the spec very selectively (hint: there are strong and weak comparion functions).

stokito commented:
Yeah, that's why I said "in fact should do". This is not a big deal while Tomcat generates those ETag itself. I'm not even sure that there is any clients who sends multiple ETags. But the change improves performance a lot.

reschke commented:
> But the change improves performance a lot.

Can you quantify that?

stokito commented:
At least twice fast because and creates less memory garbage.
Here is a benchmark with results https://gist.github.com/stokito/a82eed1aef6ad965e2a279825f1c3420
I tested with force GC between tests to compare just CPU time. Note that this comparison performed for each request.

stokito commented:
I found related ticket https://bz.apache.org/bugzilla/show_bug.cgi?id=64265

reschke commented:
> At least twice fast because creates less memory garbage

And this is measurable in practice., not in an isolated unit test?

stokito commented:
Ok, so Tomcat should make a weak matching.
I changed the description of PR and added commits. Now this is a bug fix and simplification is only in last commit.
You can merge everything except of the last commit d4831ba.

Since logic of matching now even more complicated I updated the benchmark:
https://gist.github.com/stokito/a82eed1aef6ad965e2a279825f1c3420
But more important is that it just looks simpler.

stokito commented:
I added another two commits with performance optimization. Given the nature of ETags that Tomcat generates itself I made some priority to check ETag:
1. Full match (what we normally expect)
2. Weak match
3. Asterisk match (not sure who ever used it)

stokito commented:
OMG you right, they differ:


     An origin server MUST use the strong comparison function when
     comparing entity-tags for If-Match (Section 2.3.2), since the client
     intends this precondition to prevent the method from being applied if
     there have been any changes to the representation data.

     A recipient MUST use the weak comparison function when comparing
     entity-tags for If-None-Match (Section 2.3.2), since weak entity-tags
     can be used for cache validation even if there have been changes to
     the representation data.

But what is funny is that in Tomcat on the contrary `If-Match` is *weak* while  `If-None-Match` is *strong*.

I'll fix that and update the PR

stokito commented:
@michael-o now I fixed, please review

michael-o commented:
Will do tomorrow.

michael-o commented on java/org/apache/catalina/servlets/DefaultServlet.java:
```diff
@@ -2611,6 +2574,44 @@ private PrecompressedResource(WebResource resource, CompressionFormat format) {
+    private boolean matchByEtagStrong(String headerValue, String eTag) {
+        // BZ 64265: Default servlet uses weak matching so we strip any leading "W/" and
+        // then compare using equals
+        String resourceEtag = weakEtagToStrong(eTag);
```
This does not feel right because if the ETag is weak, the comparison can never succeed.

stokito replied:
the tomcat's own Weak etags will be converted to strong i.e. the W/ is remored inside weakEtagToStrong(). So it's fine.
You can run a test that checks it.

stokito commented on java/org/apache/catalina/servlets/DefaultServlet.java:
```diff
@@ -2611,6 +2574,44 @@ private PrecompressedResource(WebResource resource, CompressionFormat format) {
+        // BZ 64265: Default servlet uses weak matching so we strip any leading "W/" and
+        // then compare using equals
+        String resourceEtag = weakEtagToStrong(eTag);
+        StringTokenizer commaTokenizer = new StringTokenizer(headerValue, ",");
```
yes, because contains() can't be used for the strict comparison. We can do a similar custom function (i.e. without creation of heavy StringTokenizer) but TBH the `If-Match` is used rarely and as far I understood it was intended for WebDav's PUT requests.

reschke replied:
PUT with conditional header fields is standard base HTTP (yes, used in WebDAV, but in other scenarios as well).

stokito commented on java/org/apache/catalina/servlets/DefaultServlet.java:
```diff
@@ -2611,6 +2574,44 @@ private PrecompressedResource(WebResource resource, CompressionFormat format) {
+            return true;
+        }
+        // asterisk checked last as rarely used
+        return headerValue.equals("*");
```
it is cheap but rare in practice. Browsers never sends asterisk, it is only needed for some tricky WebDav put requests.

michael-o replied:
That's right, we use it for updates via REST. But still, the implementation should be complete.

stokito replied:
maybe I missed something, but this is complete with spec. The only diff is that we'll check for full ETags firts

reschke replied:
PUT with "If-None-Match: *" is standard HTTP, not "tricky WebDAV". It means: "fail the request if the request would overwrite an existing representation". (see https://greenbytes.de/tech/webdav/rfc7232.html#rfc.section.3.2.p.5)

stokito commented on java/org/apache/catalina/servlets/DefaultServlet.java:
```diff
@@ -2611,6 +2574,44 @@ private PrecompressedResource(WebResource resource, CompressionFormat format) {
+     */
+    private boolean matchByEtagWeak(String headerValue, String eTag) {
+        // Match W/"1" and W/"1"
+        if (headerValue.contains(eTag)) {
```
Yes, it may look not so obvious for those who never used such approach. Still it makes a lot of sense to use this.
In 99% (if not all 100) cases the If-None-Match will be just a single ETag that Tomcat generated itself.
But create an instance of the `StringTokenizer` allocates 160 bytes, and each a new `nextToken()` call takes about 88 bytes for etag like `W/"1047-1578315296666"`.
The contains() call have no any allocations and it is optimized intrinsic function in JVM

reschke replied:
I *believe* just doing "contains()" will yield incorrect results if the field value is malformed, such as `WW/"1"`.

stokito replied:
GIGO

markt-asf replied:
Garbage in, garbage out is not an acceptable way to handle HTTP headers. Failure to reject malformed HTTP headers from clients can lead to security issues - typically request smuggling - when a proxy takes a different approaching to allowing the invalid header to the back-end server although in this instance the proxy would need to do something pretty unusual.
The right solution here is to implement RFC 7232 compliant parsing of `entity-tag` in `org.apache.tomcat.util.http.parser`
Tomcat has been tightening up the parsing of HTTP headers over time, generally improving things as parsing issues are raised with each header. It appears that now is the time to address `entity-tag`. I also note that we should make the use of weak comparison for `If-Match` configurable. Users that extend the Default servlet to provide strong ETags (or use a custom resource implementation) will almost certainly want to use a strong comparison here.
I'll start work on a suitable parser.

markt-asf commented:
Simplifying the matching is the wrong approach. The matching needs to be tightened up to be more spec compliant. A change to that effect has now been implemented.
